### PR TITLE
docs: use theming submodule, update docs build and linking

### LINKED ISSUE #1517: update docs check to use current tket dev version
Currently the docs check is not always using the newest state of pytket in the repo, this should be updated.

See https://github.com/CQCL/tket/pull/1516

https://github.com/CQCL/tket/pull/1512#issuecomment-2265182248

### PR DESCRIPTION
**Pull request by CalMacCQ**
# Description

Updating the API docs build to use `pytket-docs-theming` as a submodule. This is already in place for the extensions.

Also no longer using a pypi version of pytket to build the docs in C.I. (rather a develop wheel). fixes  #1517 

This means using a shared `conf.py` file so I've deleted that along with the `_static` dir.

I've also updated the `display.rst` page to use the MyST-NB library for showing code snippets. This ensures visual consistency with other parts of the TKET docs.

I've cleaned up some of the crosslinking to the extension packages in `extensions.md`.

Also updated all links to point to docs.quantinuum.com/tket rather than tket.quantinuum.com

# Related issues

closes  #1517

### CODE REVIEW

File: nix-support/pytket.nix [hunk 1]
```diff
@@ -1,13 +1,6 @@
+{ package_version }:
 self: super:
 let
-  config_contents = builtins.readFile ../pytket/docs/conf.py;
-  versions =
-    builtins.match ''.*release *= *["']([^"']+)["'].*'' config_contents;
```
Comment (CalMacCQ) [inline nix-support/pytket.nix:5]: Changeed by Jake in c3f417b1a1356352d97cded6dd57cf87d2368e29

File: pytket/docs/extensions.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+```
+pip install pytket-quantinuum
+```
+
+The types of `Backend` available in pytket are the following
+
+## Types of Backend
+
+- **QPUs** - These are real quantum computers that return shots based results. E.g the [QuantinuumBackend](inv:#*.QuantinuumBackend) .
+- **Cloud Access** - Cloud backends allow pytket to interface with cloud platforms to access additional QPUs and simulators. E.g [BraketBackend](inv:#*.BraketBackend) .
```
Comment (CalMacCQ) [inline pytket/docs/extensions.md:19]: Updated some of these links to use robust intersphinx linking. Will do the rest in another PR.

File: pytket/docs/circuit.rst [hunk 3]
```diff
@@ -7,7 +7,7 @@ pytket.circuit
    circuit_class.rst
    optype.rst
    classical.rst
-   display.rst
+   display.md
```
Comment (cqc-alec) [inline pytket/docs/circuit.rst:10]: Are you planning to change the others to markdown as well?

File: pytket/docs/circuit_class.rst [hunk 4]
```diff
@@ -338,7 +338,7 @@ condition on a specified set of bit values.)
 
    .. Note:: For adding boxes to a circuit the :py:meth:`Circuit.add_gate` method is sufficient to append any :py:class:`OpType` to a :py:class:`Circuit`.
 
-   .. jupyter-input::
+   ::
```
Comment (cqc-alec) [inline pytket/docs/circuit_class.rst:341]: What effect does this have?

**Review by cqc-alec — COMMENTED** [review]
There are a lot of error in the logs from the docs build [here](https://github.com/CQCL/tket/actions/runs/11329451816/job/31505039606?pr=1609). Are these expected?

**Comment by CalMacCQ** [comment]
> There are a lot of error in the logs from the docs build [here](https://github.com/CQCL/tket/actions/runs/11329451816/job/31505039606?pr=1609). Are these expected?

Ah this is a problem. Thanks for pointing this out. The C.I. build was still passing but there are still some problems as it allows warnings

File: pytket/docs/circuit_class.rst (hunk 4 above)
Comment (CalMacCQ) [reply pytket/docs/circuit_class.rst:341]: None really. It just doesn't use the jupyter sphinx package.

It'll still show as a code snippet

File: .github/workflows/build_and_test.yml [hunk 5]
```diff
@@ -312,12 +312,11 @@ jobs:
       if: github.event_name == 'pull_request' || github.event_name == 'workflow_dispatch'
       timeout-minutes: 20
       run: |
-        pip install poetry==1.6.1
+        pip install poetry==1.8.1
+        poetry config virtualenvs.create false
```
Comment (CalMacCQ) [inline .github/workflows/build_and_test.yml:316]: Here I'm not using the default `.venv` with poetry and instead using the editable wheel installed earlier in the workflow

File: pytket/docs/extensions.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+## Cloud Access
+
+[BraketBackend](inv:#*.BraketBackend)
+\- Interface to Amazon Braket service.
+
+## Emulators
+
+[IBMQEmulatorBackend](inv:#*.IBMQEmulatorBackend) - A backend which uses the [AerBackend](inv:#*.AerBackend) to emulate the behavior of IBMQBackend.
+
+[QuantinuumBackend](inv:#*.QuantinuumBackend)
```
Comment (CalMacCQ) [inline pytket/docs/extensions.md:54]: Note that there are some additional build warnings given by this new syntax. I think these are due to the fact that `QuantinuumBackend` (and other backends) don't have a unique import path.

The links do work however and I've ran a link checker over these pages to fix any broken links.

I'll make an issue to fix the warnings. The output is quite cluttered already

File: pytket/docs/circuit.rst (hunk 3 above)
Comment (CalMacCQ) [reply pytket/docs/circuit.rst:10]: Not in a rush to do that.

I updated `display.md` because I wanted to use the MyST-NB library to show the code snippets on that page (consistent styling for the rest of the docs)

I updated the extensions index page because I wanted to use the more robust linking syntax where I don't have to use the full URL.

There are automated tools for doing the conversion so I could possibly update the other files in a subsequent PR.

File: pytket/docs/extensions.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+[QulacsBackend](inv:#*.QulacsBackend)
+\- Backend for running simulations of variational quantum circuits on the Qulacs simulator.
+
+```{toctree}
+:caption: 'Extensions:'
+:maxdepth: 0
+
+pytket-aqt <https://cqcl.github.io/pytket-aqt/api/index.html>
+pytket-azure <https://docs.quantinuum.com/tket/extensions/pytket-azure>
+pytket-braket <https://docs.quantinuum.com/tket/extensions/pytket-braket>
```
Comment (CalMacCQ) [inline pytket/docs/extensions.md:115]: Unfortunately the linking syntax doesn't straightforwardly work with these TOCtree links which is annoying. I'll have to come back to this.

File: .github/workflows/build_and_test.yml [hunk 8]
```diff
@@ -312,18 +312,17 @@ jobs:
...
-        poetry run -C pytket/docs ./.github/workflows/build-docs
-        
+        bash ./build-docs.sh
     - name: Upload artifact
       if: github.event_name == 'pull_request' || github.event_name == 'workflow_dispatch'
       uses: actions/upload-artifact@v4
       with:
         name: pytket_docs
-        path: pytket/docs/build/html/
+        path: pytket/docs/build/
```
Comment (CalMacCQ) [inline .github/workflows/build_and_test.yml:325]: This is so that the artifact goes to the right place

File: .github/workflows/build_and_test.yml [hunk 9]
```diff
@@ -312,18 +312,17 @@ jobs:
       if: github.event_name == 'pull_request' || github.event_name == 'workflow_dispatch'
       timeout-minutes: 20
       run: |
-        pip install poetry==1.6.1
+        pip install poetry==1.8.1
```
Comment (cqc-alec) [inline .github/workflows/build_and_test.yml:315]: Is it important that it be exactly this version of `poetry` or can we just assume the latest version will work?

File: pytket/docs/extensions.md (hunk 6 above)
Comment (cqc-alec) [reply pytket/docs/extensions.md:54]: I can just see one warning that might be important, but it can be fixed in another PR (or this one if you prefer):
```
/home/runner/work/tket/tket/pytket/pytket/utils/operators.py:docstring of pytket.utils.QubitPauliOperator.all_qubits:7: WARNING: Definition list ends without a blank line; unexpected unindent.
```

File: pytket/docs/install.rst [hunk 10]
```diff
@@ -28,7 +28,7 @@ Building TKET from source
 
 TKET can be built from source by compiling the C++. This is now possible on MacOS, Windows and Linux (including ARM Linux).
 
-For instructions on how to do this see the `tket repository README <https://github.com/CQCL/tket#how-to-build-tket-and-pytket>`_. 
+For instructions on how to do this see the `tket repository README <https://github.com/CQCL/tket?tab=readme-ov-file#how-to-build-tket-and-pytket>`_.
```
[the remark this answers is not in the document]
Comment (CalMacCQ) [reply pytket/docs/install.rst:31]: Link checker was complaining about this for some reason. However as you point out it worked fine so I'll revert back to the orginal link.

File: pytket/docs/backends.rst [hunk 11]
```diff
@@ -1,13 +1,13 @@
...
+Contains :py:class:`Backend` abstract class and associated methods. In pytket a :py:class:`~pytket.backends.Backend` represents an interface between pytket and a quantum device or simulator. Different backends are defined in the various pytket extension modules and inherit from the core pytket :py:class:`~pytket.backends.Backend` class.
 
-There are several `example notebooks <https://tket.quantinuum.com/examples>`_ on pytket :py:class:`Backend`\s. If you are interested in developing your own :py:class:`Backend` or pytket extension then see the `creating backends <https://tket.quantinuum.com/user-guide/examples/backends/creating_backends.html>`_ tutorial.
+There are several example notebooks on pytket :py:class:`~pytket.backends.Backend`\s. If you are interested in developing your own :py:class:`Backend` or pytket extension then see the `creating backends <https://docs.quantinuum.com/tket/user-guide/examples/backends/creating_backends.html>`_ tutorial.
```
[the remark this answers is not in the document]
Comment (CalMacCQ) [reply pytket/docs/backends.rst:6]: This was actually an invalid link. However I agree I should remove the setnence as well.

File: .github/workflows/build_and_test.yml (hunk 9 above)
Comment (CalMacCQ) [reply .github/workflows/build_and_test.yml:315]: I updated it so that I could use `package-mode=false` in `pyproject.toml` this clarifies that we're only using poetry to install the docs dependencies. We are not using it for package management here.

Seems this isn't available in 1.6.1.

File: pytket/docs/index.rst [hunk 12]
```diff
@@ -48,7 +48,7 @@ To install the ``pytket-quantinuum`` package use the following command.
     pip install pytket-quantinuum
 
 The extensions supported by tket are described
-`here <https://tket.quantinuum.com/api-docs/extensions>`_.
+`here <extensions.html>`_.
```
[the remark this answers is not in the document]
Comment (CalMacCQ) [reply pytket/docs/index.rst:51]: Gone with .html for both for consistency

File: pytket/docs/install.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (CalMacCQ) [reply pytket/docs/install.rst]: removed [523aa2a](https://github.com/CQCL/tket/pull/1609/commits/523aa2aab30900f1c6cfc70841947f7799338ea9)

File: .github/workflows/build_and_test.yml (hunk 9 above)
Comment (cqc-alec) [reply .github/workflows/build_and_test.yml:315]: So we can probably make it `>=1.8.1`? Less likely to have dependency conflicts in the future if we do that.

File: pytket/docs/extensions.md (hunk 6 above)
Comment (CalMacCQ) [reply pytket/docs/extensions.md:54]: I'd rather fix that last warning in a follow up PR if thats okay. Not immediately sure what the matter is.

File: pytket/docs/backends.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (CalMacCQ) [reply pytket/docs/backends.rst]: removed [523aa2a](https://github.com/CQCL/tket/pull/1609/commits/523aa2aab30900f1c6cfc70841947f7799338ea9)

File: pytket/docs/install.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (CalMacCQ) [reply pytket/docs/install.rst]: used the original -> [6859a1e](https://github.com/CQCL/tket/pull/1609/commits/6859a1e52b8a11312ce23e87029c71c31a8add16)

File: .github/workflows/build_and_test.yml (hunk 9 above)
Comment (CalMacCQ) [reply .github/workflows/build_and_test.yml]: Good idea, updated [dd03ddd](https://github.com/CQCL/tket/pull/1609/commits/dd03dddb24a9a0ee65982c1c3fa5656c97ff085c)
